Resetting dump statistics after event

cedrichansen opened the pull request:
- When a dump is printed, the hook statistics are reset
- Changing the ms to us to reflect change in OMR.
- Dependant on omr change: https://github.com/eclipse/omr/pull/4221

Doc issue https://github.com/eclipse/openj9-docs/issues/364

Signed-off-by: Cedric Hansen <cedric.hansen@ibm.com>

dmitripivkine commented:
@LinHu2016 Would you please review?

LinHu2016 commented:
LGTM

dmitripivkine commented:
@amicic FYI

amicic commented:
Jenkins compile win jdk11

keithc-ca commented on runtime/rasdump/javadump.cpp:
```diff
@@ -2021,7 +2021,7 @@ JavaCoreDumpWriter::writeHookInfo(struct OMRHookInfo4Dump *hookInfo)
 	_OutputStream.writeInteger64(hookInfo->startTime % 1000, ".%03llu");
 	_OutputStream.writeCharacters("\n");
-	_OutputStream.writeCharacters("4HKDURATION           DurationMs: ");
+	_OutputStream.writeCharacters("4HKDURATION           Durationμs: ");
```
I don't think we want non-ASCII content in javacore files.

amicic replied:
@cedrichansen please, fix. we commonly use plain 'u' instead.

cedrichansen replied:
will do!

keithc-ca commented on runtime/rasdump/javadump.cpp:
```diff
@@ -2051,6 +2056,20 @@ JavaCoreDumpWriter::writeHookInterface(struct J9HookInterface **hookInterface)
 			_OutputStream.writeCharacters("NULL\n");
 		}
+		//reset the eventDump statistics
+		eventDump->count = 0;
```
This represents a change in behavior that not all users will want. What's the motivation for resetting those statistics?

amicic replied:
Motivation.... Relevant (typically, relatively recent) stats may get 'polluted' with irrelevant old data from long runs. Especially true for longestHook stats and to some extent for totalTime (and average which can be deduced). 

Backward compatibilitiy..... Yes, it's change of behavior. But, if we reset, and we have all reports/dumps available, we can reconstruct those numbers for the whole run, if really needed.

Affected users.... This is relatively new feature and rarely used, and I believe not documented. Typically, by us, designers, and Service team. We consulted @manqingl from Service and he was ok with it.

manqingl replied:
Yes, I requested this change from service perspective.

smithwil commented:
I would support having a note in the javacore saying that these statistics are measured since the previous javacore. Does the reset apply to all subsections of the “HOOK subcomponent dump routine”? It should be consistent across all subsections

keithc-ca commented:
I suggest it would be better to present both kinds of statistics: from the begging and from the most recent dump. That way users can pay attention to the information they want and don't have to reconstruct anything.

amicic commented:
I think it's reasonable to add a note in java core itself, at the start of the report/section, perhaps? @cedrichansen, please add it.

I believe reset apply to all subsections, but I'm not familiar with all the code. @LinHu2016, please confirm.

I'm against maintaining both, due to runtime overhead. They are mostly updated atomically, plus there is a bit of footprint increase.

manqingl commented:
I recall the option to maintain both was discussed, and it was suggested to not to keep both.  The major argument was that the very first javacore would have the value from the beginning.  I do not see any obvious reason to combine the times in all javacore files.

dmitripivkine commented:
Another reason is to exclude JVM startup time from hooks usage picture.

LinHu2016 commented:
> I would support having a note in the javacore saying that these statistics are measured since the previous javacore.
> 
> Does the reset apply to all subsections of the “HOOK subcomponent dump routine”? It should be consistent across all subsections

yes, the reset apply to all subsections of HOOK dump routine.

LinHu2016 commented:
> I think it's reasonable to add a note in java core itself, at the start of the report/section, perhaps? @cedrichansen, please add it.
> 
> I believe reset applies to all subsections, but I'm not familiar with all the code. @LinHu2016, please confirm.
> 
> I'm against maintaining both, due to runtime overhead. They are mostly updated atomically, plus there is a bit of footprint increase.

I agree with Aleks that keeping both would cause double footprint usage for storing the stats, it is too much overhead comparing the benefit.

keithc-ca commented:
While it's true that keeping both sets of statistics will (at most) double the footprint, I don't agree it is excessive.

The doesn't need to be any additional runtime overhead if we manage it properly. The data collection need not change: only the action of generating a javacore file would be affected - it would update the second set of statistics so information from the start and from the last javacore can be presented. The 'reset' action would copy counts to the since-last-javacore, update maximum duration, reset current maximum duration.

pshipton commented:
The HOOKS section of the javacore is documented https://www.eclipse.org/openj9/docs/dump_javadump/
When the final behavior is decided, please open an Issue at https://github.com/eclipse/openj9-docs to reflect the change.

keithc-ca commented on runtime/rasdump/javadump.cpp:
```diff
@@ -2043,6 +2043,11 @@ JavaCoreDumpWriter::writeHookInterface(struct J9HookInterface **hookInterface)
+
+			_OutputStream.writeCharacters("3HKTOTALTIME       ");
+			_OutputStream.writeInteger(eventDump->totalTime, "%zu");
+			_OutputStream.writeCharacters("μs\n"); 
```
This 'μ' needs to be ASCII as well.

keithc-ca commented on runtime/rasdump/javadump.cpp:
```diff
@@ -2051,6 +2056,20 @@ JavaCoreDumpWriter::writeHookInterface(struct J9HookInterface **hookInterface)
 			}
 			_OutputStream.writeCharacters("NULL\n");
 		}
+		//reset the eventDump statistics
```
Please use C style comments (`/* ... */`) like (most of) the rest of this file.

keithc-ca commented:
I think we can start with this change and add the second set of statistics if users request it.

We do need a message in the javacore file that clarifies that the data only represent the period since the previous javacore.

LinHu2016 commented:
> The HOOKS section of the javacore is documented https://www.eclipse.org/openj9/docs/dump_javadump/
> When the final behavior is decided, please open an Issue at https://github.com/eclipse/openj9-docs to reflect the change.

https://github.com/eclipse/openj9-docs/issues/364

cedrichansen commented:
I have made a few light modifications to this PR.
First, I added a note in the HOOK subcomponent dump routine that indicates that the statistics only represent the period since the previous javacore.
Second, I modified the remaining "μ"'s into regular u.
Lastly, modified my comment to conform to the comments in most of the rest of the file.

Any additional feedback is welcome!

keithc-ca commented on runtime/rasdump/javadump.cpp:
```diff
@@ -2524,7 +2543,7 @@ void
 {
 	/* Write the section header */
-	_OutputStream.writeCharacters("0SECTION       HOOK subcomponent dump routine\n");
+	_OutputStream.writeCharacters("0SECTION       HOOK subcomponent dump routine NOTE: This data only represents the period since the previous javacore\n");
```
I think this note should appear on a separate line from the `0SECTION`; javacore files are purposely structured to make parsing easier - this might negate some of that.

cedrichansen replied:
Would adding a line right below the "0SECTION       HOOK subcomponent dump routine" line, saying something like "0NOTE         This data only represents the period since the previous javacore" , be an reasonable fix?

keithc-ca replied:
That sounds good, but I would put it below the section separator, beginning with `1NOTE`:
```
NULL           ===============================
1NOTE          This data only represents the period since the previous javacore
```

smithwil commented:
For the note I would prefer something like "This data is reset every time a javacore is taken."

Saying "This data only represents the period since the previous javacore" implies that there was a previous javacore, but that is not true for the first javacore from a process. It would be good to avoid concerns that the complaints that the "previous" javacore was not produced, not reported in logs, or has been lost somehow.

keithc-ca commented on runtime/rasdump/javadump.cpp:
```diff
@@ -2021,7 +2021,7 @@ JavaCoreDumpWriter::writeHookInfo(struct OMRHookInfo4Dump *hookInfo)
 	_OutputStream.writeInteger64(hookInfo->startTime % 1000, ".%03llu");
 	_OutputStream.writeCharacters("\n");
-	_OutputStream.writeCharacters("4HKDURATION           DurationMs: ");
+	_OutputStream.writeCharacters("4HKDURATION           Durationus: ");
```
I would prefer the presentation adopted below for total time (with 'us' after the value).

cedrichansen replied:
Agreed. The addition has been made

keithc-ca replied:
Please remove the (now) redundant 'us' suffix from 'Durationus', but keep the length of that string the same by adding two more spaces after the colon(`:`).

keithc-ca replied:
I had asked that you put those extra spaces *after* the colon, not before, but I won't hold this up for that.

keithc-ca commented on runtime/rasdump/javadump.cpp:
```diff
@@ -2526,7 +2545,7 @@ JavaCoreDumpWriter::writeHookSection(void)
 	_OutputStream.writeCharacters("0SECTION       HOOK subcomponent dump routine\n");
 	_OutputStream.writeCharacters("NULL           ==============================\n");
-
+	_OutputStream.writeCharacters("1NOTE		  This data is reset every time a javacore is taken\n");
```
Please don't embed tabs within string literals: use spaces for consistent alignment.

cedrichansen replied:
My mistake. The string now contains only spaces

keithc-ca commented on runtime/rasdump/javadump.cpp:
```diff
@@ -2526,7 +2545,7 @@ JavaCoreDumpWriter::writeHookSection(void)
 	_OutputStream.writeCharacters("0SECTION       HOOK subcomponent dump routine\n");
 	_OutputStream.writeCharacters("NULL           ==============================\n");
-
+	_OutputStream.writeCharacters("1NOTE          This data is reset every time a javacore is taken\n");
```
Grammatical nit: 'data' is plural so it ought to say 'These data are reset...' or something else that's consistent.

cedrichansen commented:
@keithc-ca The 2 changes you have requested have been made. Let me know if anything else needs to be changed

keithc-ca commented:
Jenkins test sanity plinux,win32 jdk8

keithc-ca commented:
Restarting builds due to #7287.

keithc-ca commented:
Jenkins test sanity plinux,win32 jdk8

pshipton commented:
@cedrichansen please cherry pick your comment and create a PR against the v0.17.0-release branch.
